attestation/golang/golang.go

```diff
@@ -0,0 +1,396 @@
+	}
+
+	for path, product := range products {
+		for _, mimeType := range mimeTypes {
+			if !strings.Contains(mimeType, product.MimeType) {
+				continue
+			}
+		}
+
+		newDigestSet, err := cryptoutil.CalculateDigestSetFromFile(path, ctx.Hashes())
+		if newDigestSet == nil || err != nil {
+			return fmt.Errorf("error calculating digest set from file: %s", path)
+		}
+
+		if !newDigestSet.Equal(product.Digest) {
+			return fmt.Errorf("integrity error: product digest set does not match candidate digest set")
+		}
+
+		f, err := os.Open(path)
+		if err != nil {
+			return fmt.Errorf("error opening file: %s", path)
+		}
+		defer f.Close()
+
+		scanner := bufio.NewScanner(f)
+
+		a.Packages = map[string]Package{}
+		a.CoverageEnabled = false
+		var cancel bool
+
+		for scanner.Scan() {
+			line := scanner.Bytes()
+
+			var output GoOutput
+			err := json.Unmarshal(line, &output)
+			if err != nil {
+				log.Debugf("(attestation/golang) error unmarshaling go output file: %w", err)
+				// NOTE: we want to move to the next product
+				cancel = true
+				break
+			}
+
+			var ok bool
+			var pack Package
+			if pack, ok = a.Packages[output.Package]; !ok {
+				pack = Package{
+					Element: Element{
+						Name:    output.Package,
+						Outputs: []string{},
+					},
+					Tests: map[string]Test{},
+				}
+			}
+
+			if output.Test != "" {
+				var test Test
+				if test, ok = pack.Tests[output.Test]; !ok {
+					test = Test{
+						Element: Element{
+							Name:    output.Package,
+							Outputs: []string{},
+						},
+					}
+				}
+
+				var tmpPercent float64 = 0
+				parseJsonBlock(&test.Element, &tmpPercent, output)
+				if tmpPercent != 0 {
+					log.Debugf("(attestation/golang) unexpected percentage %f found in output %s for test %s", tmpPercent, output.Output, output.Test)
+				}
+
+				pack.Tests[output.Test] = test
+			} else {
+				foundPercent := parseJsonBlock(&pack.Element, &pack.PercentageCoverage, output)
+				if foundPercent {
+					a.CoverageEnabled = foundPercent
+				}
+			}
+
+			a.Packages[output.Package] = pack
+		}
+
+		if cancel {
+			continue
+		}
+
+		if len(a.ExcludedPackages) > 0 {
+			for _, name := range a.ExcludedPackages {
+				if pack, ok := a.Packages[name]; ok {
+					if len(pack.Tests) > 0 {
+						log.Warnf("(attestation/golang) package %s containing tests excluded from attestation", name)
+					}
+					delete(a.Packages, name)
+				} else {
+					log.Warnf("(attestation/golang) excluded package %s not found in test results", name)
+				}
+			}
+		}
+
+		totalPass := true
+		totalCoverage := 0.0
+		for _, p := range a.Packages {
+			if p.PercentageCoverage != 0 {
+				totalCoverage += p.PercentageCoverage
+			}
+
+			// NOTE: we only need to check the total package's test for a pass/fail
+			if !p.Pass {
+				totalPass = false
+			}
+		}
+
+		// NOTE: to get the average we need to divide by the number of packages
+		totalCoverage = totalCoverage / float64(len(a.Packages))
```

Comment:
If a package is empty would it divide by zero?
Maybe we can avoid it
```suggestion
		if len(a.Packages) > 0 {
				totalCoverage = totalCoverage / float64(len(a.Packages))
		} else {
    				totalCoverage = 0
		}
```